Add scalars method to connection and session classes

miguelgrinberg opened the pull request:
Add scalars() method to connection and session classes.

Fixes #6990 

### Description
This change adds a scalars() method to the connection and session classes that combines the execution of the query with the conversion to scalars. A stream_scalars() method is also added to the asyncio connection and session classes to apply the same shortcut to streaming queries.

### Checklist
<!-- go over following points. check them with an `x` if they do apply, (they turn into clickable checkboxes once the PR is submitted, so no need to do everything at once)

-->

This pull request is:

- [ ] A documentation / typographical error fix
	- Good to go, no issue or tests are needed
- [ ] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [x] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**

miguelgrinberg committed:
Add scalars method to connection and session classes

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 9b798edfc09f1704f791173999f0f844588b1e0c of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
New Gerrit review created for change 9b798edfc09f1704f791173999f0f844588b1e0c: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3065

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision a421c9c9d22a8d189651443c028f2e5a7a60cde9 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset a421c9c9d22a8d189651443c028f2e5a7a60cde9 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3065

CaselIT commented:
I really do need to add async tests here on github, the current pipelines only test the sync sqlite

miguelgrinberg commented:
@CaselIT Heh, yeah! I assumed the aiosqlite tests were running with the base set and didn't bother to check. I even congratulated myself for doing such a great job with these tests which passed the first time. :)

I believe these tests should be okay now. :)

CaselIT commented:
> I believe these tests should be okay now. :)

Indeed. 
pep8 has failed on gerrit:
```
pep8 run-test: commands[0] | flake8 ./lib/ ./test/ ./examples/ setup.py doc/build/conf.py
pep8 run-test: commands[1] | black --check ./lib/ ./test/ ./examples/ setup.py doc/build/conf.py
would reformat lib/sqlalchemy/ext/asyncio/engine.py
would reformat test/orm/test_session.py
```

It would be great if you could also add a changelog file in `doc/build/changelog/unreleased_14`

miguelgrinberg commented:
@CaselIT Okay, 🤞  that this is the one. I am not very familiar with the tags that you are using, I looked at a few of the entries in the change log and entered what seemed appropriate, but let me know if you have better tags to suggest. Thanks!

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 3122e4a4e2fc86dade12c6c99d652e311ba83678 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset 3122e4a4e2fc86dade12c6c99d652e311ba83678 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3065

sqla-tester commented on lib/sqlalchemy/orm/session.py:
```diff
@@ -1724,6 +1724,31 @@ def scalar(
+
+        Usage and parameters are the same as that of
+        :meth:`_orm.Session.execute`; the return result is a
+        :class:`_result.ScalarResult` filtering object which
```
**Federico Caselli** (CaselIT) wrote:

I'm not an authority on English,  but filter here seems to imply that some rows are filtered out. Maybe it's just me :).
An alternative off the top of my head "yielding objects" ?

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3065

miguelgrinberg replied:
I agree that "filtering" can be interpreted as a selection process instead of a conversion, but if we are going to change it then it should be changed in a few more places that currently use this terminology in existing code. For example, [here is one](https://github.com/sqlalchemy/sqlalchemy/blob/680baea6b83c497e862f3f0ade28e11940cfd895/lib/sqlalchemy/engine/result.py#L857).

The word "mapping" seems like it is more appropriate than filtering here, but mapping can also make things confusing in the context of the ORM.

CaselIT replied:
I had not noticed it was already used there.



> The word "mapping" seems like it is more appropriate than filtering here, but mapping can also make things confusing in the context of the ORM.

I agree that it would fit, but it may indeed clash with orm mapping.

Maybe a general "transform"?

sqla-tester replied:
**mike bayer** (zzzeek) wrote:

lets do "transform" in a separate fix since I used "filtering" already.  i believe this might be internal terminology.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3065

sqla-tester reviewed (commented):
**Federico Caselli** (CaselIT) wrote:

> Patch Set 2:
> 
> miguelgrinberg@github wrote:
> 
> @CaselIT Okay, 🤞  that this is the one. I am not very familiar with the tags that you are using, I looked at a few of the entries in the change log and entered what seemed appropriate, but let me know if you have better tags to suggest. Thanks!

looks fine to me

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3065

sqla-tester commented on lib/sqlalchemy/ext/asyncio/session.py:
```diff
@@ -231,6 +231,31 @@ async def scalar(
+        bind_arguments=None,
+        **kw
+    ):
+        """Execute a statement and return scalar results.
```
**mike bayer** (zzzeek) wrote:

all the new methods need to have a version note:

    .. versionadded:: 1.4.24

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3065

sqla-tester reviewed (commented):
**mike bayer** (zzzeek) wrote:

all the docstrings need to have a "..versionadded:: 1.4.24" directive.   if we dont make it in 1.4.24 we can mass update later.

also, we need a quick test in test/engine/test_execute.py->ExecuteTest for the connection-level scalars() method, which can also dedicate a test to the scalar() method too.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3065

miguelgrinberg commented:
Added the version number to the docstrings, and the two new unit tests that you requested.

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 0c7ee32cd4c54737905e6375309bb9436a559f30 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset 0c7ee32cd4c54737905e6375309bb9436a559f30 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3065

sqla-tester commented:
**mike bayer** (zzzeek) wrote:

that failure is a false positive, this is probably ready, will look

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3065

sqla-tester commented:
**mike bayer** (zzzeek) wrote:

oop it's some oracle thing.  i want to get this in so ill update

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3065

miguelgrinberg commented:
@zzzeek This was my own mistake, I used an incorrect decorator in the test, which made the test run for Oracle instead of Postgres, sqlite, etc. But I have corrected it two days ago. Do you need to manually start the full CI build? Because it did not run again after I fixed this mistake on the PR.

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision f5bf0796ffda2a1824a663ae0b4b87112b3431ab of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset f5bf0796ffda2a1824a663ae0b4b87112b3431ab added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3065

CaselIT commented:
yes, I was missing the last version on gerrit

sqla-tester commented:
**mike bayer** (zzzeek) wrote:

> Patch Set 4:
> 
> miguelgrinberg@github wrote:
> 
> @zzzeek This was my own mistake, I used an incorrect decorator in the test, which made the test run for Oracle instead of Postgres, sqlite, etc. But I have corrected it two days ago. Do you need to manually start the full CI build? Because it did not run again after I fixed this mistake on the PR.

i have review comments here of things to change, if you can work on these today sometime we can get them on the next round.  i want to work towards a 1.4.24 release this week, like wednesady ish if possible.   i will publish review comments momentarily


View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3065

sqla-tester commented on doc/build/changelog/unreleased_14/6990.rst:
```diff
@@ -0,0 +1,8 @@
+    :tags: usecase, engine, orm
+    :tickets: 6990
+
+    This change adds a scalars() method to the connection and session classes
```
**mike bayer** (zzzeek) wrote:

I have new text for this:

    Added new methods :meth:`_orm.Session.scalars`,
    :meth:`_engine.Connection.scalars`, :meth:`_asyncio.AsyncSession.scalars`
    and :meth:`_asyncio.AsyncSession.stream_scalars`, which provide a short cut
    to the use case of receiving a row-oriented :class:`_result.Result` object
    and converting it to a :class:`_result.ScalarResult` object via the
    :meth:`_engine.Result.scalars` method, to return a list of values rather
    than a list of rows. The new methods are analogous to the long existing
    :meth:`_orm.Session.scalar` and :meth:`_engine.Connection.scalar` methods
    used to return a single value from the first row only. Pull request
    courtesy Miguel Grinberg.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3065

sqla-tester commented on lib/sqlalchemy/engine/base.py:
```diff
@@ -1161,6 +1161,17 @@ def scalar(self, object_, *multiparams, **params):
         return self.execute(object_, *multiparams, **params).scalar()
 
+    def scalars(self, object_, *multiparams, **params):
+        """Executes and returns the first column of each row.
```
**mike bayer** (zzzeek) wrote:

"executes and returns a scalar result set"

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3065

sqla-tester commented on lib/sqlalchemy/engine/base.py:
```diff
@@ -1161,6 +1161,17 @@ def scalar(self, object_, *multiparams, **params):
+    def scalars(self, object_, *multiparams, **params):
+        """Executes and returns the first column of each row.
+
+        The underlying result/cursor is closed after execution.
```
**mike bayer** (zzzeek) wrote:

this line is not correct here, this now returns a ScalarResult that acts as a filter around the CursorResult.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3065

sqla-tester commented on lib/sqlalchemy/engine/base.py:
```diff
@@ -1161,6 +1161,17 @@ def scalar(self, object_, *multiparams, **params):
+        """Executes and returns the first column of each row.
+
+        The underlying result/cursor is closed after execution.
+
```
**mike bayer** (zzzeek) wrote:

let's also add the `:return:` for all new methods, in this case:

   :return: a :class:`_result.ScalarResult`

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3065

sqla-tester commented on lib/sqlalchemy/ext/asyncio/engine.py:
```diff
@@ -439,6 +439,48 @@ async def scalar(
+
+        """
+        result = await self.execute(statement, parameters, execution_options)
+        print("+++", result)
```
**mike bayer** (zzzeek) wrote:

remove print() statement

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3065

sqla-tester commented on lib/sqlalchemy/ext/asyncio/engine.py:
```diff
@@ -439,6 +439,48 @@ async def scalar(
+        parameters=None,
+        execution_options=util.EMPTY_DICT,
+    ):
+        r"""Executes a SQL statement construct and returns a stream of scalar
```
**mike bayer** (zzzeek) wrote:

"executes a SQL statement and returns a streaming scalar result object"

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3065

sqla-tester commented on lib/sqlalchemy/ext/asyncio/session.py:
```diff
@@ -231,6 +231,33 @@ async def scalar(
+        **kw
+    ):
+        """Execute a statement and return scalar results.
+
```
**mike bayer** (zzzeek) wrote:

add

    :return: an :class:`_result.ScalarResult` object

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3065

sqla-tester commented on lib/sqlalchemy/ext/asyncio/session.py:
```diff
@@ -287,6 +314,33 @@ async def stream(
+        **kw
+    ):
+        """Execute a statement and return a stream of scalar results.
+
```
**mike bayer** (zzzeek) wrote:

add

    :return: an :class:`_asyncio.AsyncScalarResult` object

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3065

sqla-tester commented on lib/sqlalchemy/orm/session.py:
```diff
@@ -1724,6 +1724,33 @@ def scalar(
+        :meth:`_orm.Session.execute`; the return result is a
+        :class:`_result.ScalarResult` filtering object which
+        will return single elements rather than :class:`_row.Row` objects.
+
```
**mike bayer** (zzzeek) wrote:

add

    :return:  a :class:`_result.ScalarResult` object

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3065

sqla-tester commented on test/engine/test_execute.py:
```diff
@@ -707,6 +707,36 @@ def test_works_after_dispose_testing_engine(self):
+    def test_scalar(self, connection):
+        conn = connection
+        conn.exec_driver_sql(
+            "insert into users (user_id, user_name) " "values (?, ?)",
```
**mike bayer** (zzzeek) wrote:

the "?" is actually not named paramstyle.    For these tests we dont have to deal with any of that explicitly.     These two tests can be written as follows:


    def test_scalar(self, connection):
        conn = connection
        users = self.tables.users
        conn.execute(
            users.insert(),
            [
                {"user_id": 1, "user_name": "sandy"},
                {"user_id": 2, "user_name": "spongebob"},
            ],
        )
        res = conn.scalar(
            select(users.c.user_name).order_by(users.c.user_id)
        )
        eq_(res, "sandy")

    def test_scalars(self, connection):
        conn = connection
        users = self.tables.users
        conn.execute(
            users.insert(),
            [
                {"user_id": 1, "user_name": "sandy"},
                {"user_id": 2, "user_name": "spongebob"},
            ],
        )
        res = conn.scalars(
            select(users.c.user_name).order_by(users.c.user_id)
        )
        eq_(res.all(), ["sandy", "spongebob"])

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3065

sqla-tester reviewed (commented):
**mike bayer** (zzzeek) wrote:

OK here's a list of things to change

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3065

miguelgrinberg commented:
All changes applied. 🤞

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision b3e0bb3042c55b0cc5af6a25cb3f31b929f88a47 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset b3e0bb3042c55b0cc5af6a25cb3f31b929f88a47 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3065

sqla-tester commented:
**mike bayer** (zzzeek) wrote:

thanks for the effort and patience on this!

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3065

sqla-tester commented:
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3065 has been **merged**. Congratulations! :)
